common/transport/amqp/src/amqp.ts

```diff
@@ -311,19 +551,52 @@ export class Amqp {
         },
         disconnecting: {
           _onEnter: (disconnectCallback, err) => {
-            const disconnect = (callback) => {
-              if (err) {
+            debug('Entering disconnecting state with disconnectCallback: ' + disconnectCallback + ' error of: ' + this._getErrorName(err));
+            const sessionEnd = (callback) => {
+              //
+              // If a disconnection has already happened then there is no point in trying to send a session close.
+              // Just be done.
+              //
+              if (this._disconnectionOccurred) {
                 callback();
               } else {
-                /*Codes_SRS_NODE_COMMON_AMQP_16_004: [The disconnect method shall call the done callback when the application/service has been successfully disconnected from the service] */
-                this._amqp.disconnect().then(() => {
-                  debug('amqp10 client cleanly disconnected');
+                //
+                // A session close may have already been received from the peer.  If we are disconnecting because of a session error as an example.
+                // We should send a session close from our end BUT, we should not expect to receive back another session close in response.
+                //
+                this._rheaSession.close();
+                if (this._sessionCloseOccurred) {
                   callback();
-                  return null;
-                }).catch((err) => {
-                  debug('amqp10 failed to cleanly disconnect: ' + err.toString());
+                } else {
+                  this._sessionCallback = callback; // So that the session_close handler for this disconnecting state can invoke the callback.
+                }
+              }
+            };
```

Comment:
this tells me "disconnecting" needs to be broken down into 2 states at least (for session and connection).

Reply:
I'm assuming that it's not simply the }; that is signaling you here.

Really couldn't disagree more.  These are fairly intertwined concepts, for instance the session code is going to have to deal with the fact that the connection has already been disconnected where for the happy path you should be able to do the session stuff before doing the connection stuff.  Doesn't work that way.  A boatload of cases have come through this code.   Trying to separate them will only produce very repetitive code and LESS systematic treatment.

---
In reply to: [209133207](https://github.com/Azure/azure-iot-sdk-node/pull/344#discussion_r209133207) [](ancestors = 209133207)